### PR TITLE
adding multi-with-statements support

### LINKED ISSUE #116: --py39-plus: multi-with-statements
Add support for trailing commas for this:

```diff
 with (
     foo as bar,
-    baz as womp
+    baz as womp,
 ):
     # ...
```

### PR DESCRIPTION
**Pull request by keller00**
Fixes #116

### CODE REVIEW

File: add_trailing_comma/_plugins/with.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+from add_trailing_comma._ast_helpers import ast_to_offset
+from add_trailing_comma._data import register
+from add_trailing_comma._data import State
+from add_trailing_comma._data import TokenFunc
+from add_trailing_comma._plugins.imports import _find_import as _find_cm
+from add_trailing_comma._token_helpers import fix_brace
+
+
+if sys.version_info >= (3, 9):  # pragma: >=3.9 cover
```
Comment (asottile) [inline add_trailing_comma/_plugins/with.py:19]: this shouldn't be like this -- it should be `--py39-plus`

File: add_trailing_comma/_plugins/with.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+from typing import Iterable
+
+from tokenize_rt import Offset
+from tokenize_rt import Token
+
+from add_trailing_comma._ast_helpers import ast_to_offset
+from add_trailing_comma._data import register
+from add_trailing_comma._data import State
+from add_trailing_comma._data import TokenFunc
+from add_trailing_comma._plugins.imports import _find_import as _find_cm
```
Comment (asottile) [inline add_trailing_comma/_plugins/with.py:15]: aliasing this unrelated function seems wrong -- also plugins generally shouldn't import from other plugins.  if there's common code it should get extracted

File: add_trailing_comma/_plugins/with.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
+from __future__ import annotations
```
Comment (asottile) [inline add_trailing_comma/_plugins/with.py:1]: this file has an invalid module name

File: add_trailing_comma/_plugins/with.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        i: int,
+        tokens: list[Token],
+        add_comma: bool = False,
+    ) -> None:
+        fix_brace(
+            tokens,
+            _find_cm(i, tokens),
+            add_comma=add_comma,
+            remove_comma=True,
+        )
```
Comment (asottile) [inline add_trailing_comma/_plugins/with.py:30]: this doesn't work -- here's an example:

```python
with (
     open("wat") 
) as f, open("2") as f2: pass
```

File: tests/features/with_test.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+from add_trailing_comma._main import _fix_src
+from add_trailing_comma._main import main
+
+
+@pytest.mark.parametrize(
+    'src',
+    (
+        # Simple context manager inside parentheses
+        'from threading import Lock\n'
+        'with(Lock()):\n'
```
Comment (asottile) [inline tests/features/with_test.py:16]: I would never write code like this -- put a space after the `with` keyword

File: add_trailing_comma/_plugins/_with.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            )  # pragma: >=3.9 cover
+
+
+@register(ast.With)
+def visit_With(
+        state: State,
+        node: ast.With,
+) -> Iterable[tuple[Offset, TokenFunc]]:
+    multiwith = len(node.items) > 1
+    if state.min_version >= (3, 9) and multiwith:
```
Comment (asottile) [inline add_trailing_comma/_plugins/_with.py:57]: no reason to assign this

File: add_trailing_comma/_plugins/_with.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        brace_stack = [first_brace]
+
+        for i in range(first_brace + 1, len(tokens)):
+            token = tokens[i]
+            if token.src in START_BRACES:
+                brace_stack.append(i)
+            elif token.src in END_BRACES:
+                brace_stack.pop()
+            if len(brace_stack) == 1 and token.name == 'NL':
+                multiline_withitems = True
```
Comment (asottile) [inline add_trailing_comma/_plugins/_with.py:35]: why the stack check here?

File: tests/features/with_test.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+from add_trailing_comma._main import _fix_src
+from add_trailing_comma._main import main
+
+
+@pytest.mark.parametrize(
+    ('src', 'plus39'),
+    (
+        pytest.param(
+            'from threading import Lock\n'
+            'with(Lock() as l):\n'
```
Comment (asottile) [inline tests/features/with_test.py:16]: none of these tests look like how I would expect someone to write code -- it's useful to have one test without whitespace after the `with` token but unless there's a good reason most of the tests should have a space after `with` (it's not a function after all)

File: tests/features/with_test.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+            '    pass',
+            False,
+            id='simple named cm',
```
Comment (asottile) [inline tests/features/with_test.py:19]: what is `cm` (I know what it is, but there's no reason to abbreviate)

File: tests/features/with_test.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+import sys
+
+import pytest
+
+from add_trailing_comma._main import _fix_src
+from add_trailing_comma._main import main
+
+
+@pytest.mark.parametrize(
+    ('src', 'plus39'),
```
Comment (asottile) [inline tests/features/with_test.py:12]: this is a [boolean trap](https://youtu.be/CnRkXO_a5mI) -- either take the actual version or make it a separate test function

File: tests/features/with_test.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+            'with (open("/tmp/t.py") as t, open("/tmp/y.py") as y): pass',
+            True,
+            id='single line',
+        ),
+        pytest.param(
+            'with (\n'
+            '    open("wat")\n'
+            ') as f, open("2") as f2: pass',
+            True,
+            id='unusual example',
```
Comment (asottile) [inline tests/features/with_test.py:44]: don't just say "unusual", explain what's unusual about it

File: add_trailing_comma/_main.py [hunk 12]
```diff
@@ -94,6 +94,10 @@ def main(argv: Sequence[str] | None = None) -> int:
         '--py36-plus',
         action='store_const', dest='min_version', const=(3, 6),
     )
+    parser.add_argument(
+        '--py39-plus',
+        action='store_const', dest='min_version', const=(3, 9),
+    )
```
Comment (asottile) [inline add_trailing_comma/_main.py:100]: actually now that I look at it, I can't think of a situation where this flag actually helps.  if they're using the python3.9+ syntax already that's the only situation it'd rewrite I think ?  maybe the original way was better with the version check in the module similar to `Match`

Comment (keller00) [reply add_trailing_comma/_main.py:100]: I think so too :+1:

File: add_trailing_comma/_plugins/_with.py (hunk 7 above)
Comment (keller00) [reply add_trailing_comma/_plugins/_with.py:35]: My idea was basically to not rewrite parentheses based on internal line break
```
with (open(
    "a",
    some_other_really_long_parameter=True,
) as a, self.lock: pass
```
TBH this example isn't perfect as parenthesized expression will be broken the next line which would make the with statement have a new line in it at the top level.
Would love to know your thoughts on this.

Comment (asottile) [reply add_trailing_comma/_plugins/_with.py:35]: isn't the `:` check below going to cover that?

Comment (asottile) [reply add_trailing_comma/_plugins/_with.py:35]: and when you've got weird cases like that -- include them in tests!

Comment (keller00) [reply add_trailing_comma/_plugins/_with.py:35]: I made a typo, sorry!
```python
with (open(
    "a",
    some_other_really_long_parameter=True,
) as a, self.lock): pass```

Comment (asottile) [reply add_trailing_comma/_plugins/_with.py:35]: I would expect it to come out like this

```python
with (
    open(
        "a",
        some_other_really_long_parameter=True,
    ) as a, self.lock,
): pass
```

and you branch _does_ do that -- but it takes 2 passes to get there

**Review by asottile — APPROVED** [review]
I ended up simplifying it a bunch -- thanks for working on this !

**Pull request merged**